### PR TITLE
Feature/enhance tests

### LINKED ISSUE #66: Add Warning if Peak_load_window is used with v2g
Right now the code doesn't check if v2g is set to true. Therefore one doesn't notice that v2g is not supported. At least a warning should be given.

### PR DESCRIPTION
**Pull request by Piranias**
Fix #66

**Changes proposed in this pull request**:
- add test folder and tests for strategies and the generate files
- still needed: input files for "generate_from_download" and "generate_from_simbev" (see #67)
- open question: "kann es sein, dass greedy_market nur ohne v2g funktioniert? In meinen Tests läuft nur das scenario ohne v2g durch, in der Doku steht aber v2g supported. Oder ist das Teil der v2g-Bugs die grad noch bearbeitet werden?" - wurde aus den tests rausgenommen, wird hier ( #69) weiter bearbeitet
- add warning to peak_load_window if v2g is used (#66)
- add tests for general scenario outputs
- add tests explicitly for flex_window outputs
 

The following steps were realized, as well (required):
- [x] Correct linting with `fake8 path_to_script`
- [x] Check if tests pass locally (`pytest ./src/tests.py`)

Also the following steps were realized (if applies):
- [ ] Write docstrings to your code
- [ ] For new functionalities: Explain in readthedocs
- [x] Write test(s) for your new patch of code

*Please mark above checkboxes as following:*
- [ ] Open
- [x] Done

### CODE REVIEW

File: examples/generate.cfg [hunk 1]
```diff
@@ -3,9 +3,9 @@
 # python generate.py --config examples/generate.cfg
 
 # set output file
-output = examples/scenario.json
+output = examples/scenario_B.json
```
Comment (stefansc1) [inline examples/generate.cfg:6]: Why is this changed? scenario_B does not seem to exist

File: examples/generate.cfg [hunk 2]
```diff
@@ -3,9 +3,9 @@
...
 # set vehicle types file
-vehicle_types = examples/vehicle_types.json
+vehicle_types = vehicle_types.json
```
Comment (stefansc1) [inline examples/generate.cfg:8]: This is relative to your working directory. As the executable files are in the root folder, this means you'd need the default vehicle_types file in the root folder as well (which I'm against, as it should only serve as an example). Please revert.

File: generate.py [hunk 3]
```diff
@@ -102,7 +105,7 @@ def generate(args):
     ext = args.vehicle_types.split('.')[-1]
     if ext != "json":
         print("File extension mismatch: vehicle type file should be .json")
-    with open(args.vehicle_types) as f:
+    with open(path.join(target_path, args.vehicle_types)) as f:
```
Comment (stefansc1) [inline generate.py:108]: The vehicle_types file is only used for generating the file, NOT in the simulation. The "all paths are relative to the target path" is referring to the paths in the scenario.json, not all paths during generation as well. The vehicle_types path is given as absolute or relative to the working directory - irrespective of the target_path.
Having the files needed by the simulation as relative to the scenario file location just makes it easier to find them again during the simulation (which might have a different working directory), without having to resort to absolute paths. The path to the vehicle_types file is not saved in the simulation file, so there is no need to adapt the path.
For clarity's sake, you could change the comment in line 80 to `file paths in the output file are relative to that file`.

File: generate.py [hunk 4]
```diff
@@ -400,7 +401,8 @@ def generate(args):
     parser.add_argument('--include-price-csv-option', '-po', metavar=('KEY', 'VALUE'),
                         nargs=2, default=[], action='append',
                         help='append additional argument to price signals')
-    parser.add_argument('--config', help='Use config file to set arguments')
+    parser.add_argument('--config', help='Use config file to set arguments',
+                        default="examples/generate.cfg")
```
Comment (stefansc1) [inline generate.py:405]: I don't like that. The config file is a major change to all arguments. It should only be set actively, not because this argument is not set. Your change makes it much easier to unwittingly use the config, much harder to not use a config at all and therefore more likely to generate unexpected results (more so, as you are not warned when overriding arguments by providing a config file). It is cleaner to set the necessary options yourself instead of relying on the content of a file that might have been changed. Use the argparse Namespace functionality, there is a reason why parsing the arguments and the functionality are separated.

File: generate_from_csv.py [hunk 5]
```diff
@@ -21,6 +21,9 @@ def generate_from_csv(args):
     if missing:
         raise SystemExit("The following arguments are required: {}".format(", ".join(missing)))
 
+    # all paths are relative to output file
+    target_path = path.dirname(args.output)
+
```
Comment (stefansc1) [inline generate_from_csv.py:26]: see generate.py

File: generate_from_csv.py [hunk 6]
```diff
@@ -34,7 +37,7 @@ def generate_from_csv(args):
     ext = args.vehicle_types.split('.')[-1]
     if ext != "json":
         print("File extension mismatch: vehicle type file should be .json")
-    with open(args.vehicle_types) as f:
+    with open(path.join(target_path, args.vehicle_types)) as f:
```
Comment (stefansc1) [inline generate_from_csv.py:40]: see generate.py

File: src/strategies/peak_load_window.py [hunk 7]
```diff
@@ -83,6 +84,10 @@ def step(self, event_list=[]):
         standing = {}
         energy_needed = 0
         for vid, v in self.world_state.vehicles.items():
+            if v.vehicle_type.v2g:
+                warnings.warn("The vehicle type of vehicle {} is set to v2g although v2g is not "
+                              "supported with peak_load_window strategy. V2g will thus be "
+                              "neglected.".format(vid))
```
Comment (stefansc1) [inline src/strategies/peak_load_window.py:90]: I would prefer the capitalization of V2G (instead of lower-case letters), as it is an abbreviation.

**Review by stefansc1 — CHANGES_REQUESTED** [review]
More general comments:

* don't work with config files for your tests, but provide the necessary arguments in an argparse Namespace. This forces you to think about which parameters are really necessary. The test config files are just copies from the example files and not minimal. You can even reuse the Namespace for different tests, so you don't always have to parse a new config file.
* there is a typo in your test_generate.py file name
* test_util is also part of test_components. I am fine with splitting the file, since test_strategies has been moved as well.

File: generate.py (hunk 4 above)
Comment (Piranias) [reply generate.py:405]: ok!

File: generate.py [hunk 8]
```diff
@@ -380,7 +380,6 @@ def generate(args):
     parser.add_argument('--gc-power', type=int, default=530, help='set power at grid connection '
                                                                   'point in kW')
     parser.add_argument('--seed', default=None, type=int, help='set random seed')
-
```
Comment (stefansc1) [inline generate.py:383]: This empty line was in there to separate the block with more specific arguments for generate and the file arguments. It doesn't hurt, makes it a bit more structured and this is the only change in this file. I'd say revert this change, so there is one file less in the diff.

File: generate_from_csv.py [hunk 9]
```diff
@@ -151,6 +151,7 @@ def generate_from_csv(args):
     # save path and options for CSV timeseries
     # all paths are relative to output file
     target_path = path.dirname(args.output)
+
```
Comment (stefansc1) [inline generate_from_csv.py:154]: Similar argumentation as for generate.py, this is the only change in this file.

File: src/scenario.py [hunk 10]
```diff
@@ -640,3 +642,9 @@ def run(self, strategy_name, options):
                 plt.setp(ax.get_xticklabels(), rotation=30, ha='right')
 
             plt.show()
+
+        ###########################
+        # get some numbers for testing
+        max_total_load = max(totalLoad) if totalLoad else 0
+
+        return(max_total_load)
```
Comment (stefansc1) [inline src/scenario.py:650]: I would avoid giving that function a return value. Proposed change: make it a class attribute `self.max_total_load = ...`
In addition, totalLoad is always set (initialized as empty array), so the check is not needed.

File: tests/test_components.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+from src import battery, loading_curve, scenario
+
+
+def get_test_json():
+    # get minimum working json example
+    return {
+        "scenario": {
+            "start_time": "2020-01-01T00:00:00+02:00",
+            "interval": 15,
+            "n_intervals": 35040
```
Comment (stefansc1) [inline tests/test_components.py:14]: What is your opinion, do we still need to simulate a full year?

File: tests/test_components.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+            "vehicle_events": [],
+        },
+        "strategy": "greedy"
+    }
+
+
+class TestScenario(unittest.TestCase):
+
+    def test_scenario_times(self):
+        # corect number of timesteps?
```
Comment (stefansc1) [inline tests/test_components.py:34]: oops, typo in `correct`

File: tests/test_data/input_test_generate/vehicle_types.json [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    },
+    "all_information": {
+        "name": "Displayed name",
+        "capacity": 0,
+        "mileage": 0,
+        "charging_curve": [[0,1],[1,0]],
+        "min_charging_power": 0,
+        "v2g": false,
+        "battery_efficiency": 0.95
+    }
```
Comment (stefansc1) [inline tests/test_data/input_test_generate/vehicle_types.json:47]: Are these three vehicle_types needed? I did not find these in the test_generate.

File: tests/test_components.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        "events": {
+            "external_loads": {},
+            "grid_operator_signals": [],
+            "vehicle_events": [],
+        },
+        "strategy": "greedy"
+    }
+
+
+class TestScenario(unittest.TestCase):
```
Comment (stefansc1) [inline tests/test_components.py:31]: Duplicate of test_scenarios.py

File: src/scenario.py [hunk 15]
```diff
@@ -537,34 +537,36 @@ def run(self, strategy_name, options):
...
+            for k, v in step.items():
+                if k not in loads:
+                    # new key, not present before
+                    loads[k] = [0] * i
+                loads[k].append(v)
+            for k in loads.keys():
+                if k not in step:
+                    # old key not in current step
+                    loads[k].append(0)
+
```
Comment (stefansc1) [inline src/scenario.py:564]: Why is this outside of the visual check? It does not seem to be used if you're not planning to plot. In the test_scenario file, I found only max_total_load to be used.

**Review by stefansc1 — CHANGES_REQUESTED** [review]

File: generate_from_csv.py (hunk 9 above)
Comment (Piranias) [reply generate_from_csv.py:154]: so you like the line here or not? When changing back the target_path I liked a blank line here. But I guess it does not really matter much

Comment (stefansc1) [reply generate_from_csv.py:154]: No, I don't like the newline. It disrupts the initialization block and is the only change to this file, making the diff bigger as it needs to be.

File: src/scenario.py (hunk 10 above)
Comment (Piranias) [reply src/scenario.py:650]: The check is in case the totalLoad is an empty list, in this case we cannot get the max of it. 

We could also add it as a class attrribute. But I think we will need to add some more parameters in order to test various scenarios/strategies, so quite alot more class attributes, that we only need for testing. Do you still like this better?

File: tests/test_components.py (hunk 11 above)
Comment (Piranias) [reply tests/test_components.py:14]: I don't see a reason why this is necessary. For what reason did you set it up like this?

File: tests/test_data/input_test_generate/vehicle_types.json (hunk 13 above)
Comment (Piranias) [reply tests/test_data/input_test_generate/vehicle_types.json:47]: true, i will clear it out

File: src/scenario.py (hunk 10 above)
Comment (stefansc1) [reply src/scenario.py:650]: You're right, I was thinking about `sum()`. For `max()`, this makes sense.
Regarding the return value: for many testing parameters, I would still prefer to save them as class attributes, even more so. Maybe they could be clustered into a testing or meta or whatever attribute.

File: tests/test_components.py (hunk 11 above)
Comment (stefansc1) [reply tests/test_components.py:14]: Seeing as nothing actually happens (no events), it does not really matter how long we simulate. For a minimal test, this could also be set to one day.

File: src/util.py [hunk 16]
```diff
@@ -21,8 +21,9 @@ def datetime_from_isoformat(s):
 
     # fallback: use strptime. Problem is timezone with colon
     # Thanks SO! (https://stackoverflow.com/questions/30999230/how-to-parse-timezone-with-colon)
-    if s[-3:-2] == ':':
-        s = s[:-3]+s[-2:]
+    if s[-6:-5] == "+":
```
Comment (stefansc1) [inline src/util.py:24]: datetimes in SpiceEV should always be timezone-aware. Not sure if we should support timezone-agnostic parsing in an outdated Python version. Anyway, this can be simplified:
```
if s[-6] == "+" and s[-3] == ":":
    s = s[:-3] + s[-2:]
```

File: tests/test_generate.py [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    "--battery": [[350, 0.5]],
+    "--start-time": '2018-01-01T00:15:00+00:00',
+    "--vehicle_types": "test_data/input_test_generate/vehicle_types.json",
+    "--include-ext-load-csv": None,
+    "--include-ext-csv-option": [],
+    "--include-feed-in-csv": None,
+    "--include-feed-in-csv-option": [],
+    "--seed": None,
+    "--include_price_csv": None,
+    "--include_price_csv_option": []
```
Comment (stefansc1) [inline tests/test_generate.py:32]: These can be omitted, as these are the default values.

File: tests/test_strategies.py [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+
+    def test_greedy_B(self):
+        input = os.path.join(TEST_REPO_PATH, 'test_data/input_test_strategies/scenario_B.json')
+        s = scenario.Scenario(load_json(input), os.path.dirname(input))
+        s.run('greedy', {})
+
+    def test_balanced_B(self):
+        input = os.path.join(TEST_REPO_PATH, 'test_data/input_test_strategies/scenario_B.json')
+        s = scenario.Scenario(load_json(input), os.path.dirname(input))
+        s.run('balanced', {})
```
Comment (stefansc1) [inline tests/test_strategies.py:103]: Is there something that can be tested along the way, e.g., if balanced allocates the same energy for every timestep?

**Review by stefansc1 — CHANGES_REQUESTED** [review]

File: tests/test_strategies.py (hunk 18 above)
Comment (Piranias) [reply tests/test_strategies.py:103]: yes, probably. I only added some specific tests for flex_window, but more should and will follow for the other strategies.

**Review by stefansc1 — APPROVED** [review]
